### PR TITLE
Hard: Kava delegators earn HARD rewards via the Incentive module

### PR DESCRIPTION
**Pull request by denalimarsh**
Move delegator rewards backed by Staking module hooks into the Incentive module.

TODO:
- [x] tests

### CODE REVIEW

File: x/incentive/keeper/rewards.go [hunk 1]
```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+	if !found {
+		previousRewardFactor = sdk.ZeroDec()
+	}
+	newRewardFactor := previousRewardFactor.Add(rewardFactor)
+	k.SetHardDelegatorRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
+	k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+	return nil
+}
+
+// TODO: could a single delegation update the delegator index, effectively updating the value before
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:512]: Not sure what scenario this is referring to. Can you specify it a bit more?

File: x/incentive/keeper/rewards.go [hunk 2]
```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+	userRewardFactor := claim.DelegatorRewardIndexes[delegatorIndex].RewardFactor
+	rewardsAccumulatedFactor := delagatorFactor.Sub(userRewardFactor)
+	if rewardsAccumulatedFactor.IsZero() {
+		return
+	}
+	claim.DelegatorRewardIndexes[delegatorIndex].RewardFactor = delagatorFactor
+
+	var totalDelegated sdk.Dec
+
+	// TODO: set reasonable max limit on delegation iteration
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:431]: Good point. Worth looking at an export of kava and cosmos and seeing the max number of delegations that any address has. My guess is we don't want to iterate over more than ~100 in a transaction. Good to document this number when we choose it.

File: x/incentive/keeper/rewards.go [hunk 3]
```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+		// Delegators don't accumulate rewards if their validator is unbonded/slashed
+		if validator.GetStatus() != sdk.Bonded {
+			continue
+		}
+
+		if validator.GetTokens().IsZero() {
+			continue
+		}
+
+		sharesToTokens := sdk.NewDecFromInt(validator.GetTokens()).Quo(validator.GetDelegatorShares())
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:449]: to get `delegatedTokens`,I believe you can do `validator.TokensFromShares(delegation.GetShares())`

see: https://github.com/cosmos/cosmos-sdk/blob/2727c14183d05f5c31ca11f2dcd52b0aa683fadf/x/staking/exported/exported.go#L33

File: x/incentive/keeper/rewards.go [hunk 4]
```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+		delegatedTokens := sharesToTokens.Mul(delegation.GetShares())
+		if delegatedTokens.IsZero() || delegatedTokens.IsNegative() {
+			continue
+		}
+
+		totalDelegated = totalDelegated.Add(delegatedTokens)
+		// TODO: do we need to consider respective share of validator tokens?
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:456]: don't think so. Just calculating the number of bonded tokens that this delegator accounts for is sufficient.

File: x/incentive/keeper/rewards.go [hunk 5]
```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+	}
+
+	// Add rewards to delegator's hard claim
+	newRewardsCoin := sdk.NewCoin(types.HardLiquidityRewardDenom, rewardsEarned)
+	claim.Reward = claim.Reward.Add(newRewardsCoin)
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// AccumulateHardDelegatorRewards updates the rewards accumulated for the input reward period
+func (k Keeper) AccumulateHardDelegatorRewards(ctx sdk.Context, delAddr sdk.AccAddress, rewardPeriod types.RewardPeriod) error {
```
Comment (karzak) [inline x/incentive/keeper/rewards.go:473]: Accumulate method looks good

**Review by karzak — CHANGES_REQUESTED** [review]
Approach looks good

File: x/incentive/keeper/rewards.go (hunk 4 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:454]: Removed TODO in https://github.com/Kava-Labs/kava/pull/776/commits/4ec5fe20e9d8cde0a46fda55dcf02e42e3afd732

File: x/incentive/keeper/rewards.go (hunk 3 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go:447]: Nice, thanks. Updated in https://github.com/Kava-Labs/kava/pull/776/commits/4ec5fe20e9d8cde0a46fda55dcf02e42e3afd732

File: x/incentive/keeper/hooks.go [hunk 6]
```diff
@@ -58,3 +65,58 @@ func (h Hooks) BeforeBorrowModified(ctx sdk.Context, borrow hardtypes.Borrow) {
...
+func (h Hooks) BeforeDelegationCreated(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) {
+	h.k.InitializeHardDelegatorReward(ctx, delAddr)
+}
+
+// BeforeDelegationSharesModified runs before an existing delegation is modified
+func (h Hooks) BeforeDelegationSharesModified(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) {
+	h.k.SynchronizeHardDelegatorRewards(ctx, delAddr)
+}
+
+// TODO: Consider calling the following hooks when validators are slashed/unbonded/rebonded in order
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/hooks.go]: Removed TODO in https://github.com/Kava-Labs/kava/pull/776/commits/4ec5fe20e9d8cde0a46fda55dcf02e42e3afd732

File: x/incentive/keeper/rewards.go [hunk 7]
```diff
@@ -400,6 +402,136 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+	k.SetHardDelegatorRewardFactor(ctx, rewardPeriod.CollateralType, newRewardFactor)
+	k.SetPreviousDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+	return nil
+}
+
+// TODO: could a single delegation update the delegator index, effectively updating the value before
+//		 the delegator has a chance to earn rewards on their delegations?
+// InitializeHardDelegatorReward initializes the delegator reward index of a hard claim
+func (k Keeper) InitializeHardDelegatorReward(ctx sdk.Context, delegator sdk.AccAddress) {
+	delegatorFactor, foundDelegatorFactor := k.GetHardDelegatorRewardFactor(ctx, UKAVA_DENOM)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Updated in https://github.com/Kava-Labs/kava/pull/776/commits/4ec5fe20e9d8cde0a46fda55dcf02e42e3afd732

File: x/incentive/keeper/rewards.go (hunk 1 above)
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Ah this TODO is outdated, removed in https://github.com/Kava-Labs/kava/pull/776/commits/3e33ae0354f7e0e1f658db8c9fa6e38bf5e1dc91.

File: x/incentive/keeper/hooks.go [hunk 8]
```diff
@@ -58,3 +65,47 @@ func (h Hooks) BeforeBorrowModified(ctx sdk.Context, borrow hardtypes.Borrow) {
 func (h Hooks) AfterBorrowModified(ctx sdk.Context, borrow hardtypes.Borrow) {
 	h.k.UpdateHardBorrowIndexDenoms(ctx, borrow)
 }
+
+// ------------------- Staking Module Hooks -------------------
+// TODO: how to ensure that existing delegators get their rewards?
```
Comment (karzak) [inline x/incentive/keeper/hooks.go:70]: This should be covered in migrations - existing delegator reward claims will be up to date, they will just need to be migrated from `x/hard` to `x/incentive` and initialized with a claim and a `DelegatorRewardFactor`

**Comment by karzak** [comment]
> Good point. Worth looking at an export of kava and cosmos and seeing the max number of delegations that any address has. My guess is we don't want to iterate over more than ~100 in a transaction. Good to document this number when we choose it.  

As a follow up to this, ran the numbers on latest cosmos-hub: There are 4 delegators with over 50 delegations, 17 delegators with over 20 delegations, and 70 delegators with over 10 delegations. I'm fine merging with this issue open, but will create a followup task to figure out 1) what the gas cost would look like for accessing that much state, 2) what the time complexity of such transactions would be. If it looks like 1 and 2 are problematic, one solution would be to use the staking hooks to implement an index of <delegator_address>:<total_bonded_tokens> that we can access instead of iterating delegations.

**Review by karzak — APPROVED** [review]
Minor nits, LGTM after they are addressed.

File: x/incentive/keeper/rewards.go [hunk 9]
```diff
@@ -400,6 +400,129 @@ func (k Keeper) UpdateHardBorrowIndexDenoms(ctx sdk.Context, borrow hardtypes.Bo
...
+	}
+
+	totalBonded := k.stakingKeeper.TotalBondedTokens(ctx).ToDec()
+	if totalBonded.IsZero() {
+		k.SetPreviousHardDelegatorRewardAccrualTime(ctx, rewardPeriod.CollateralType, ctx.BlockTime())
+		return nil
+	}
+
+	newRewards := timeElapsed.Mul(rewardPeriod.RewardsPerSecond.Amount)
+	delegatorFactor, found := k.hardKeeper.GetDelegatorInterestFactor(ctx, rewardPeriod.CollateralType)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/rewards.go]: Addressed in https://github.com/Kava-Labs/kava/pull/776/commits/0c5c930b1843d003cd92fc4d2d6156bc0afe9a6e

File: x/incentive/keeper/hooks.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/incentive/keeper/hooks.go]: Addressed in https://github.com/Kava-Labs/kava/pull/776/commits/0c5c930b1843d003cd92fc4d2d6156bc0afe9a6e

File: x/hard/keeper/keeper.go [hunk 10]
```diff
@@ -321,6 +321,25 @@ func (k Keeper) GetSupplyInterestFactor(ctx sdk.Context, denom string) (sdk.Dec,
 	return supplyInterestFactor, true
 }
 
+// SetDelegatorInterestFactor sets the current delegator interest factor for an individual market
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply x/hard/keeper/keeper.go]: Removed the methods in https://github.com/Kava-Labs/kava/pull/776/commits/0c5c930b1843d003cd92fc4d2d6156bc0afe9a6e

**Pull request merged**